File: src/treemapper/diffctx/edges/config/docker.py
from __future__ import annotations

import re
from pathlib import Path

from ...types import Fragment, FragmentId
from ..base import EdgeBuilder, EdgeDict, build_path_to_frags, discover_files_by_refs

_DOCKERFILE_NAMES = {"dockerfile"}
_COMPOSE_NAMES = {"docker-compose.yml", "docker-compose.yaml", "compose.yml", "compose.yaml"}

_DOCKERFILE_FROM_RE = re.compile(r"^FROM\s+(\S+)", re.MULTILINE | re.IGNORECASE)
_DOCKERFILE_COPY_RE = re.compile(r"^(?:COPY|ADD)\s+(?:--[^\s]+\s+)*([^\s]+)\s+", re.MULTILINE | re.IGNORECASE)
_DOCKERFILE_ENV_RE = re.compile(r"^ENV\s+(\w+)", re.MULTILINE | re.IGNORECASE)
_DOCKERFILE_ARG_RE = re.compile(r"^ARG\s+(\w+)", re.MULTILINE | re.IGNORECASE)

_COMPOSE_BUILD_RE = re.compile(r"^\s+build:\s*['\"]?([^'\"#\n]+)", re.MULTILINE)
_COMPOSE_CONTEXT_RE = re.compile(r"^\s+context:\s*['\"]?([^'\"#\n]+)", re.MULTILINE)
_COMPOSE_DOCKERFILE_RE = re.compile(r"^\s+dockerfile:\s*['\"]?([^'\"#\n]+)", re.MULTILINE)
_COMPOSE_DEPENDS_RE = re.compile(r"depends_on:\s*\n((?:\s+-\s*\w+\s*\n)+)", re.MULTILINE)
_COMPOSE_SERVICE_RE = re.compile(r"^(\w+):\s*$", re.MULTILINE)
_COMPOSE_VOLUME_RE = re.compile(r"^\s+-\s*['\"]?([./][^:'\"\n]+):", re.MULTILINE)


def _is_dockerfile(path: Path) -> bool:
    name = path.name.lower()
    return name in _DOCKERFILE_NAMES or name.startswith("dockerfile.") or name.endswith(".dockerfile")


def _is_compose_file(path: Path) -> bool:
    return path.name.lower() in _COMPOSE_NAMES


def _strip_dot_slash(s: str) -> str:
    while s.startswith("./"):
        s = s[2:]
    return s


def _normalize_path(base_dir: Path, rel_path: str) -> Path:
    rel_path = rel_path.strip().strip("'\"")
    rel_path = _strip_dot_slash(rel_path)
    normalized = base_dir / rel_path
    if ".." in normalized.parts:
        return base_dir
    return normalized


def _collect_docker_refs(docker_files: list[Path]) -> set[str]:
    refs: set[str] = set()
    for df in docker_files:
        try:
            content = df.read_text(encoding="utf-8")
        except (OSError, UnicodeDecodeError):
            continue

        if _is_dockerfile(df):
            _collect_dockerfile_refs(content, refs)
        if _is_compose_file(df):
            _collect_compose_refs(content, refs)
    return refs


def _collect_dockerfile_refs(content: str, refs: set[str]) -> None:
    for match in _DOCKERFILE_COPY_RE.finditer(content):
        src = match.group(1)
        if not src.startswith("--") and not src.startswith("$"):
            refs.add(_strip_dot_slash(src.strip().strip("'\"")))


def _collect_compose_refs(content: str, refs: set[str]) -> None:
    for match in _COMPOSE_BUILD_RE.finditer(content):
        refs.add(_strip_dot_slash(match.group(1).strip()))
    for match in _COMPOSE_VOLUME_RE.finditer(content):
        refs.add(_strip_dot_slash(match.group(1).strip()))


class DockerEdgeBuilder(EdgeBuilder):
    weight = 0.55
    copy_weight = 0.65
    compose_weight = 0.50
    reverse_weight_factor = 0.40

    def discover_related_files(
        self,
        changed_files: list[Path],
        all_candidate_files: list[Path],
        repo_root: Path | None = None,
    ) -> list[Path]:
        docker_files = [f for f in changed_files if _is_dockerfile(f) or _is_compose_file(f)]
        if not docker_files:
            return []

        refs = _collect_docker_refs(docker_files)
        return discover_files_by_refs(refs, changed_files, all_candidate_files, repo_root)

    def build(self, fragments: list[Fragment], repo_root: Path | None = None) -> EdgeDict:
        edges: EdgeDict = {}

        dockerfiles = [f for f in fragments if _is_dockerfile(f.path)]
        compose_files = [f for f in fragments if _is_compose_file(f.path)]

        if not dockerfiles and not compose_files:
            return edges

        path_to_frags = build_path_to_frags(fragments, repo_root)

        self._build_dockerfile_edges(dockerfiles, path_to_frags, edges)
        self._build_compose_edges(compose_files, dockerfiles, path_to_frags, edges)

        return edges

    def _build_dockerfile_edges(
        self,
        dockerfiles: list[Fragment],
        path_to_frags: dict[Path, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        for df in dockerfiles:
            self._add_copy_edges(df, path_to_frags, edges)
            self._add_env_edges(df, path_to_frags, edges)

    def _add_copy_edges(self, df: Fragment, path_to_frags: dict[Path, list[FragmentId]], edges: EdgeDict) -> None:
        base_dir = df.path.parent
        for match in _DOCKERFILE_COPY_RE.finditer(df.content):
            src_path = match.group(1)
            if src_path.startswith("--") or src_path.startswith("$"):
                continue
            self._link_copy_source(df.id, base_dir, src_path, path_to_frags, edges)

    def _link_copy_source(
        self,
        df_id: FragmentId,
        base_dir: Path,
        src_path: str,
        path_to_frags: dict[Path, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        target = _normalize_path(base_dir, src_path)
        for frag_id in path_to_frags.get(target, []):
            self.add_edge(edges, df_id, frag_id, self.copy_weight)

        if "*" in src_path:
            return
        suffix = _strip_dot_slash(src_path)
        if not suffix or suffix == ".":
            return
        for p, frag_ids in path_to_frags.items():
            if str(p).endswith(suffix):
                for frag_id in frag_ids:
                    self.add_edge(edges, df_id, frag_id, self.copy_weight * 0.8)

    def _add_env_edges(self, df: Fragment, path_to_frags: dict[Path, list[FragmentId]], edges: EdgeDict) -> None:
        env_vars = set(_DOCKERFILE_ENV_RE.findall(df.content))
        arg_vars = set(_DOCKERFILE_ARG_RE.findall(df.content))
        if not (env_vars or arg_vars):
            return

        for p in path_to_frags:
            if p.suffix.lower() == ".env" or p.name.lower().startswith(".env"):
                for frag_id in path_to_frags[p]:
                    self.add_edge(edges, df.id, frag_id, self.weight)

    def _build_compose_edges(
        self,
        compose_files: list[Fragment],
        dockerfiles: list[Fragment],
        path_to_frags: dict[Path, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        for cf in compose_files:
            self._link_nearby_dockerfiles(cf, dockerfiles, edges)
            self._link_compose_build(cf, path_to_frags, edges)
            self._link_compose_context(cf, path_to_frags, edges)
            self._link_compose_volumes(cf, path_to_frags, edges)

    def _link_nearby_dockerfiles(self, cf: Fragment, dockerfiles: list[Fragment], edges: EdgeDict) -> None:
        base_dir = cf.path.parent
        for df in dockerfiles:
            if df.path.parent == base_dir or df.path.parent.parent == base_dir:
                self.add_edge(edges, cf.id, df.id, self.compose_weight)

    def _link_compose_build(self, cf: Fragment, path_to_frags: dict[Path, list[FragmentId]], edges: EdgeDict) -> None:
        base_dir = cf.path.parent
        for match in _COMPOSE_BUILD_RE.finditer(cf.content):
            build_path = match.group(1).strip()
            if not build_path or build_path.startswith("$"):
                continue
            dockerfile_path = _normalize_path(base_dir, build_path) / "Dockerfile"
            for frag_id in path_to_frags.get(dockerfile_path, []):
                self.add_edge(edges, cf.id, frag_id, self.compose_weight)

    def _link_compose_context(self, cf: Fragment, path_to_frags: dict[Path, list[FragmentId]], edges: EdgeDict) -> None:
        base_dir = cf.path.parent
        for match in _COMPOSE_CONTEXT_RE.finditer(cf.content):
            context_path = match.group(1).strip()
            if not context_path or context_path.startswith("$"):
                continue
            target_dir = _normalize_path(base_dir, context_path)
            self._link_context_files(cf.id, target_dir, path_to_frags, edges)

    def _link_context_files(
        self,
        cf_id: FragmentId,
        target_dir: Path,
        path_to_frags: dict[Path, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        for p, frag_ids in path_to_frags.items():
            try:
                if p.is_relative_to(target_dir) or target_dir.is_relative_to(p.parent):
                    for frag_id in frag_ids:
                        self.add_edge(edges, cf_id, frag_id, self.compose_weight * 0.7)
            except (ValueError, TypeError):
                continue

    def _link_compose_volumes(self, cf: Fragment, path_to_frags: dict[Path, list[FragmentId]], edges: EdgeDict) -> None:
        base_dir = cf.path.parent
        for match in _COMPOSE_VOLUME_RE.finditer(cf.content):
            vol_path = match.group(1).strip()
            if vol_path:
                target = _normalize_path(base_dir, vol_path)
                for frag_id in path_to_frags.get(target, []):
                    self.add_edge(edges, cf.id, frag_id, self.compose_weight * 0.6)

File: diffctx/src/edges/config_edges/docker.rs
use std::path::{Path, PathBuf};

use once_cell::sync::Lazy;
use regex::Regex;
use rustc_hash::{FxHashMap, FxHashSet};

use crate::types::Fragment;

use super::super::base::{self, add_edge, EdgeBuilder, FragmentIndex, link_by_name};
use super::super::EdgeDict;

const WEIGHT: f64 = 0.55;
const COPY_WEIGHT: f64 = 0.65;
const COMPOSE_WEIGHT: f64 = 0.50;
const REVERSE_FACTOR: f64 = 0.40;

static DOCKERFILE_COPY_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?mi)^(?:COPY|ADD)\s+(?:--[^\s]+\s+)*(.+)").unwrap());
static DOCKERFILE_FROM_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?mi)^FROM\s+(\S+)").unwrap());
static DOCKERFILE_ENV_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?mi)^ENV\s+(\w+)").unwrap());
static DOCKERFILE_ARG_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?mi)^ARG\s+(\w+)").unwrap());

static COMPOSE_BUILD_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r##"(?m)^\s+build:\s*['"]?([^'"#\n]+)"##).unwrap());
static COMPOSE_CONTEXT_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r##"(?m)^\s+context:\s*['"]?([^'"#\n]+)"##).unwrap());
static COMPOSE_VOLUME_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r#"(?m)^\s+-\s*['"]?([./][^:'"\n]+):"#).unwrap());

fn is_dockerfile(path: &Path) -> bool {
    let name = path
        .file_name()
        .map(|n| n.to_string_lossy().to_lowercase())
        .unwrap_or_default();
    name == "dockerfile"
        || name.starts_with("dockerfile.")
        || name.ends_with(".dockerfile")
}

fn is_compose_file(path: &Path) -> bool {
    let name = path
        .file_name()
        .map(|n| n.to_string_lossy().to_lowercase())
        .unwrap_or_default();
    name == "docker-compose.yml"
        || name == "docker-compose.yaml"
        || name == "compose.yml"
        || name == "compose.yaml"
}

fn strip_dot_slash(s: &str) -> &str {
    let mut s = s;
    while let Some(rest) = s.strip_prefix("./") {
        s = rest;
    }
    s
}

fn split_copy_sources(raw: &str) -> Vec<&str> {
    let tokens: Vec<&str> = raw.split_whitespace().collect();
    if tokens.len() < 2 {
        return vec![];
    }
    tokens[..tokens.len() - 1].to_vec()
}

fn collect_dockerfile_refs(content: &str) -> FxHashSet<String> {
    let mut refs = FxHashSet::default();

    for cap in DOCKERFILE_COPY_RE.captures_iter(content) {
        if let Some(m) = cap.get(1) {
            for src in split_copy_sources(m.as_str()) {
                if !src.starts_with("--") && !src.starts_with('$') {
                    let cleaned = strip_dot_slash(src.trim().trim_matches(|c| c == '\'' || c == '"'));
                    if !cleaned.is_empty() {
                        refs.insert(cleaned.to_string());
                    }
                }
            }
        }
    }

    refs
}

fn collect_compose_refs(content: &str) -> FxHashSet<String> {
    let mut refs = FxHashSet::default();

    for cap in COMPOSE_BUILD_RE.captures_iter(content) {
        if let Some(m) = cap.get(1) {
            let val = strip_dot_slash(m.as_str().trim());
            if !val.is_empty() {
                refs.insert(val.to_string());
            }
        }
    }

    for cap in COMPOSE_VOLUME_RE.captures_iter(content) {
        if let Some(m) = cap.get(1) {
            let val = strip_dot_slash(m.as_str().trim());
            if !val.is_empty() {
                refs.insert(val.to_string());
            }
        }
    }

    refs
}

pub struct DockerEdgeBuilder;

impl EdgeBuilder for DockerEdgeBuilder {
    fn build(&self, fragments: &[Fragment], repo_root: Option<&Path>) -> EdgeDict {
        let dockerfiles: Vec<&Fragment> = fragments
            .iter()
            .filter(|f| is_dockerfile(Path::new(f.path())))
            .collect();
        let compose_files: Vec<&Fragment> = fragments
            .iter()
            .filter(|f| is_compose_file(Path::new(f.path())))
            .collect();

        if dockerfiles.is_empty() && compose_files.is_empty() {
            return EdgeDict::default();
        }

        let idx = FragmentIndex::new(fragments, repo_root);
        let mut edges = EdgeDict::default();

        for df in &dockerfiles {
            let copy_refs = collect_dockerfile_refs(&df.content);
            for r in &copy_refs {
                link_by_name(&df.id, r, &idx, &mut edges, COPY_WEIGHT, REVERSE_FACTOR);
            }

            let has_env = DOCKERFILE_ENV_RE.is_match(&df.content);
            let has_arg = DOCKERFILE_ARG_RE.is_match(&df.content);
            if has_env || has_arg {
                for f in fragments {
                    let fpath = Path::new(f.path());
                    let fname = fpath
                        .file_name()
                        .map(|n| n.to_string_lossy().to_lowercase())
                        .unwrap_or_default();
                    let ext = fpath
                        .extension()
                        .map(|e| e.to_string_lossy().to_lowercase())
                        .unwrap_or_default();
                    if (ext == "env" || fname.starts_with(".env")) && f.id != df.id {
                        add_edge(&mut edges, &df.id, &f.id, WEIGHT, REVERSE_FACTOR);
                    }
                }
            }
        }

        for cf in &compose_files {
            let df_dir = Path::new(cf.path()).parent();
            for df in &dockerfiles {
                let dockerfile_dir = Path::new(df.path()).parent();
                if let (Some(cf_parent), Some(df_parent)) = (df_dir, dockerfile_dir) {
                    if df_parent == cf_parent
                        || df_parent.parent() == Some(cf_parent)
                    {
                        add_edge(&mut edges, &cf.id, &df.id, COMPOSE_WEIGHT, REVERSE_FACTOR);
                    }
                }
            }

            let compose_refs = collect_compose_refs(&cf.content);
            for r in &compose_refs {
                link_by_name(&cf.id, r, &idx, &mut edges, COMPOSE_WEIGHT, REVERSE_FACTOR);
            }

            for cap in COMPOSE_CONTEXT_RE.captures_iter(&cf.content) {
                if let Some(m) = cap.get(1) {
                    let context = m.as_str().trim();
                    if !context.is_empty() && !context.starts_with('$') {
                        let context_stripped = strip_dot_slash(context);
                        for f in fragments {
                            let fpath_lower = f.path().to_lowercase();
                            if fpath_lower.contains(context_stripped) && f.id != cf.id {
                                add_edge(
                                    &mut edges,
                                    &cf.id,
                                    &f.id,
                                    COMPOSE_WEIGHT * 0.7,
                                    REVERSE_FACTOR,
                                );
                            }
                        }
                    }
                }
            }

            for cap in COMPOSE_VOLUME_RE.captures_iter(&cf.content) {
                if let Some(m) = cap.get(1) {
                    let vol = strip_dot_slash(m.as_str().trim());
                    if !vol.is_empty() {
                        link_by_name(
                            &cf.id,
                            vol,
                            &idx,
                            &mut edges,
                            COMPOSE_WEIGHT * 0.6,
                            REVERSE_FACTOR,
                        );
                    }
                }
            }
        }

        edges
    }

    fn discover_related_files(
        &self,
        changed: &[PathBuf],
        candidates: &[PathBuf],
        repo_root: Option<&Path>,
        file_cache: Option<&FxHashMap<PathBuf, String>>,
    ) -> Vec<PathBuf> {
        let docker_files: Vec<&PathBuf> = changed
            .iter()
            .filter(|p| is_dockerfile(p) || is_compose_file(p))
            .collect();
        if docker_files.is_empty() {
            return vec![];
        }

        let mut refs = FxHashSet::default();

        for df in &docker_files {
            let content = match base::read_file_cached(df, file_cache) {
                Some(c) => c,
                None => continue,
            };

            if is_dockerfile(df) {
                refs.extend(collect_dockerfile_refs(&content));
            }
            if is_compose_file(df) {
                refs.extend(collect_compose_refs(&content));
            }
        }

        base::discover_files_by_refs(&refs, changed, candidates, repo_root)
    }

    fn category_label(&self) -> Option<&str> {
        Some("docker")
    }
}
